File: claimshield-ai/server/lib/policyEngine.js
// DETERMINISTIC LAYER — no AI involved.
// Matches a claim's incidentType against the POLICIES table and reports,
// for each candidate clause, whether the evidence required by that clause
// is actually present. This is pure rule evaluation — the AI reasoning
// layer only explains these results, it never changes them.

import { POLICIES } from "../data/policies.js";

const DOC_LABELS = {
  claimForm: "Claim Form",
  incidentDescription: "Incident Description",
  repairEstimate: "Repair Estimate",
  fir: "FIR / Police Report",
};

export function evaluatePolicies(claim) {
  const candidates = POLICIES.filter((p) =>
    p.conditions.includes(`incidentType:${claim.incidentType}`)
  );

  const results = candidates.map((policy) => {
    const missing = policy.requiredEvidence.filter(
      (docKey) => !claim.documents[docKey]?.present
    );
    const present = policy.requiredEvidence.filter(
      (docKey) => claim.documents[docKey]?.present
    );

    const exclusionHit =
      claim.exclusionTriggered && policy.exclusions.includes(claim.exclusionTriggered)
        ? claim.exclusionTriggered
        : null;

    let status = missing.length === 0 ? "APPLICABLE" : "EVIDENCE_INSUFFICIENT";
    if (exclusionHit) status = "EXCLUDED";

    return {
      id: policy.id,
      number: policy.number,
      title: policy.title,
      description: policy.description,
      exclusions: policy.exclusions,
      status,
      exclusionReason: exclusionHit,
      supportingEvidence: present.map((k) => DOC_LABELS[k]),
      missingEvidence: missing.map((k) => DOC_LABELS[k]),
    };
  });

  return results;
}

export function requiredDocsForIncident(incidentType) {
  const candidates = POLICIES.filter((p) =>
    p.conditions.includes(`incidentType:${incidentType}`)
  );
  const set = new Set();
  candidates.forEach((p) => p.requiredEvidence.forEach((d) => set.add(d)));
  return Array.from(set);
}

File: claimshield-ai/src/policy_engine.py
import json
from pathlib import Path


POLICIES_FILE = Path(__file__).resolve().parent.parent / "data" / "policies.json"


def load_policies():
    with open(POLICIES_FILE, "r", encoding="utf-8-sig") as file:
        return json.load(file)


def required_docs_for_incident(incident_type):
    policies = load_policies()

    for policy in policies:
        conditions = policy.get("conditions", [])

        if f"incidentType:{incident_type}" in conditions:
            return policy.get("requiredEvidence", [])

    return []


def evaluate_policies(claim):
    policies = load_policies()
    incident_type = claim.get("incidentType")

    applicable_policies = []
    excluded_policies = []

    documents = claim.get("documents", {})

    for policy in policies:
        conditions = policy.get("conditions", [])

        if f"incidentType:{incident_type}" not in conditions:
            continue

        missing_evidence = []

        for required_doc in policy.get("requiredEvidence", []):
            document = documents.get(required_doc, {})

            if document.get("present") is not True:
                missing_evidence.append(required_doc)

        exclusion_triggered = None
        claim_exclusion = claim.get("exclusionTriggered")

        if claim_exclusion:
            for exclusion in policy.get("exclusions", []):
                if exclusion == claim_exclusion:
                    exclusion_triggered = exclusion
                    break

        result = {
            "policyId": policy.get("id"),
            "policyNumber": policy.get("number"),
            "title": policy.get("title"),
            "description": policy.get("description"),
            "applicable": True,
            "requiredEvidence": policy.get("requiredEvidence", []),
            "missingEvidence": missing_evidence,
            "exclusionTriggered": exclusion_triggered,
            "exclusions": policy.get("exclusions", []),
        }

        if exclusion_triggered:
            excluded_policies.append(result)
        else:
            applicable_policies.append(result)

    return {
        "incidentType": incident_type,
        "applicablePolicies": applicable_policies,
        "excludedPolicies": excluded_policies,
    }

